fix: Enable --skip-clone-no-changes for fork PRs (#3891)

Ulminator opened the pull request:
## what

<!--
- Describe high-level what changed as a result of these commits (i.e. in plain-english, what do these changes mean?)
- Use bullet points to be concise and to the point.
-->
This PR modifies the function signature of `VCSClient.GetFileContent` and enforces that when attempting to check the `repoCfgFile` of the `HeadBranch`, the `HeadRepo` is used instead of the `BaseRepo`.

## why

<!--
- Provide the justifications for the changes (e.g. business case). 
- Describe why these changes were made (e.g. why do these commits fix the problem?)
- Use bullet points to be concise and to the point.
-->
If one is using a forked repo, the `HeadBranch` will not exist on the `BaseRepo` (or if it does, it's not the branch you actually want).

While working in a monorepo that others would fork, the disk space on the VM that Atlantis runs on quickly filled up as every PR made would be cloned. This should not happen when the `--skip-clone-no-changes` flag is used, but the logic for that currently ignores fork PRs.

## tests

- [x] I have tested my changes by ...

I have deployed my branch and validated that PRs from forked repos are not cloned if the files changed in them are outside project directories.

`make test` passed as well.

Our workflow only uses the GitHub client so the implementation of the GitLab client has not been tested.

## references

<!--
- Link to any supporting github issues or helpful documentation to add some context (e.g. stackoverflow). 
- Use `closes #123`, if this PR closes a GitHub issue `#123`
-->
- closes #3891

jamengual commented:
did you ever tried this : https://www.runatlantis.io/docs/server-configuration.html#checkout-depth

Ulminator commented:
I did not, but I was able to confirm that [hasRepoCfg](https://github.com/runatlantis/atlantis/blob/a542aa8f2015e67957c5fdb6ec994080561aa1ec/server/events/project_command_builder.go#L346C3-L346C3) is always false for fork PRs. Since that is the case a clone will always performed, which seems to contradict the docs for `--skip-clone-no-changes`.

Can you elaborate on how `--checkout-depth` would help? I believe regardless of the `--checkout-strategy/depth` there would be many `git clone` invocations that could be avoided with this PR, but maybe I am missing something.

jamengual commented:
> I did not, but I was able to confirm that [hasRepoCfg](https://github.com/runatlantis/atlantis/blob/a542aa8f2015e67957c5fdb6ec994080561aa1ec/server/events/project_command_builder.go#L346C3-L346C3) is always false for fork PRs. Since that is the case a clone will always performed, which seems to contradict the docs for `--skip-clone-no-changes`.
> 
> Can you elaborate on how `--checkout-depth` would help? I believe regardless of the `--checkout-strategy/depth` there would be many `git clone` invocations that could be avoided with this PR, but maybe I am missing something.

I thought your concern was the size of the clone and not the amount of git clone invocations

Ulminator commented:
> > I did not, but I was able to confirm that [hasRepoCfg](https://github.com/runatlantis/atlantis/blob/a542aa8f2015e67957c5fdb6ec994080561aa1ec/server/events/project_command_builder.go#L346C3-L346C3) is always false for fork PRs. Since that is the case a clone will always performed, which seems to contradict the docs for `--skip-clone-no-changes`.
> > Can you elaborate on how `--checkout-depth` would help? I believe regardless of the `--checkout-strategy/depth` there would be many `git clone` invocations that could be avoided with this PR, but maybe I am missing something.
> 
> I thought your concern was the size of the clone and not the amount of git clone invocations

The size is not an issue if `--skip-clone-no-changes` actually skips the cloning of a forked PR when there are no changes. Those invocations are just not necessary to be performed so if they can be avoided why not?

Without this change the call to `GetFileContent` searches for the feature branch on the base repo, but it exists on the head repo so it always has the 404 status from GitHub. 

Since deploying this branch yesterday, many forked PRs have skipped the cloning flow:

<details>
  <summary>Logs</summary>

```
# cat $LOG_FILE | grep clone
{"level":"info","ts":"2023-10-25T23:31:19.114Z","caller":"events/project_command_builder.go:373","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38868"}}
{"level":"info","ts":"2023-10-25T23:52:52.236Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38823"}}
{"level":"info","ts":"2023-10-25T23:59:12.414Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38868"}}
{"level":"info","ts":"2023-10-26T00:01:20.135Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38868"}}
{"level":"info","ts":"2023-10-26T01:24:41.327Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38967"}}
{"level":"info","ts":"2023-10-26T02:10:46.179Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38789"}}
{"level":"info","ts":"2023-10-26T03:59:33.838Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38876"}}
{"level":"info","ts":"2023-10-26T09:19:26.205Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38968"}}
{"level":"info","ts":"2023-10-26T13:34:54.276Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38890"}}
{"level":"info","ts":"2023-10-26T14:00:14.383Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38890"}}
{"level":"info","ts":"2023-10-26T14:01:29.206Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38681"}}
{"level":"info","ts":"2023-10-26T14:09:26.666Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38893"}}
{"level":"info","ts":"2023-10-26T14:13:53.161Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"37073"}}
{"level":"info","ts":"2023-10-26T14:14:47.840Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38893"}}
{"level":"info","ts":"2023-10-26T14:16:21.873Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38833"}}
{"level":"info","ts":"2023-10-26T14:20:15.507Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38785"}}
{"level":"info","ts":"2023-10-26T14:22:54.801Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38969"}}
{"level":"info","ts":"2023-10-26T14:31:20.756Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38893"}}
{"level":"info","ts":"2023-10-26T14:31:41.260Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38848"}}
{"level":"info","ts":"2023-10-26T14:44:20.078Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38969"}}
{"level":"info","ts":"2023-10-26T14:45:28.691Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38885"}}
{"level":"info","ts":"2023-10-26T14:54:23.871Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38970"}}
{"level":"info","ts":"2023-10-26T14:59:53.842Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38884"}}
{"level":"info","ts":"2023-10-26T15:02:12.293Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38890"}}
{"level":"info","ts":"2023-10-26T15:08:11.103Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38823"}}
{"level":"info","ts":"2023-10-26T15:09:42.291Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38897"}}
{"level":"info","ts":"2023-10-26T15:10:41.639Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38785"}}
{"level":"info","ts":"2023-10-26T15:16:42.167Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38848"}}
{"level":"info","ts":"2023-10-26T15:24:45.334Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38890"}}
{"level":"info","ts":"2023-10-26T15:25:52.467Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38526"}}
{"level":"info","ts":"2023-10-26T15:28:22.170Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38893"}}
{"level":"info","ts":"2023-10-26T15:32:44.171Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38848"}}
{"level":"info","ts":"2023-10-26T15:37:20.356Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38833"}}
{"level":"info","ts":"2023-10-26T15:38:49.340Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38681"}}
{"level":"info","ts":"2023-10-26T15:40:07.060Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38888"}}
{"level":"info","ts":"2023-10-26T15:41:36.969Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38885"}}
{"level":"info","ts":"2023-10-26T15:42:01.283Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38885"}}
{"level":"info","ts":"2023-10-26T15:44:47.232Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38885"}}
{"level":"info","ts":"2023-10-26T15:49:12.869Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38884"}}
{"level":"info","ts":"2023-10-26T15:54:27.534Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38965"}}
{"level":"info","ts":"2023-10-26T15:55:23.129Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38965"}}
{"level":"info","ts":"2023-10-26T15:58:38.310Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38965"}}
{"level":"info","ts":"2023-10-26T16:09:19.461Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38971"}}
{"level":"info","ts":"2023-10-26T16:12:09.505Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T16:16:42.809Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"37073"}}
{"level":"info","ts":"2023-10-26T16:23:11.559Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38893"}}
{"level":"info","ts":"2023-10-26T16:23:15.545Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T16:30:47.955Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38688"}}
{"level":"info","ts":"2023-10-26T16:36:04.822Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38971"}}
{"level":"info","ts":"2023-10-26T16:36:31.320Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38973"}}
{"level":"info","ts":"2023-10-26T16:36:45.927Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38896"}}
{"level":"info","ts":"2023-10-26T16:38:54.884Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38823"}}
{"level":"info","ts":"2023-10-26T16:39:32.184Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38789"}}
{"level":"info","ts":"2023-10-26T16:41:59.160Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38879"}}
{"level":"info","ts":"2023-10-26T16:49:53.400Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38855"}}
{"level":"info","ts":"2023-10-26T16:53:25.597Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T16:56:26.702Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38896"}}
{"level":"info","ts":"2023-10-26T16:59:09.054Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38971"}}
{"level":"info","ts":"2023-10-26T16:59:40.283Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38877"}}
{"level":"info","ts":"2023-10-26T17:04:29.440Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38782"}}
{"level":"info","ts":"2023-10-26T17:06:19.428Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"37073"}}
{"level":"info","ts":"2023-10-26T17:06:51.528Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38859"}}
{"level":"info","ts":"2023-10-26T17:09:21.780Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38974"}}
{"level":"info","ts":"2023-10-26T17:10:09.035Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38865"}}
{"level":"info","ts":"2023-10-26T17:10:39.161Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38856"}}
{"level":"info","ts":"2023-10-26T17:13:05.716Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T17:13:56.948Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38975"}}
{"level":"info","ts":"2023-10-26T17:14:47.165Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38976"}}
{"level":"info","ts":"2023-10-26T17:21:00.587Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38876"}}
{"level":"info","ts":"2023-10-26T17:25:20.989Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38789"}}
{"level":"info","ts":"2023-10-26T17:32:55.331Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38841"}}
{"level":"info","ts":"2023-10-26T17:35:37.934Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38734"}}
{"level":"info","ts":"2023-10-26T17:36:13.351Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38876"}}
{"level":"info","ts":"2023-10-26T17:38:33.657Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38975"}}
{"level":"info","ts":"2023-10-26T17:39:23.620Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38804"}}
{"level":"info","ts":"2023-10-26T17:48:41.063Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38896"}}
{"level":"info","ts":"2023-10-26T17:49:13.063Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38860"}}
{"level":"info","ts":"2023-10-26T17:54:31.642Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38975"}}
{"level":"info","ts":"2023-10-26T17:58:08.096Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38877"}}
{"level":"info","ts":"2023-10-26T18:01:22.958Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T18:03:00.550Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38860"}}
{"level":"info","ts":"2023-10-26T18:04:20.174Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38804"}}
{"level":"info","ts":"2023-10-26T18:05:30.875Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38966"}}
{"level":"info","ts":"2023-10-26T18:10:07.524Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38973"}}
{"level":"info","ts":"2023-10-26T18:18:20.149Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38965"}}
{"level":"info","ts":"2023-10-26T18:20:46.205Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38830"}}
{"level":"info","ts":"2023-10-26T18:21:30.371Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38977"}}
{"level":"info","ts":"2023-10-26T18:22:15.042Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38978"}}
{"level":"info","ts":"2023-10-26T18:24:09.084Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38777"}}
{"level":"info","ts":"2023-10-26T18:28:44.619Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38972"}}
{"level":"info","ts":"2023-10-26T18:31:01.891Z","caller":"events/project_command_builder.go:369","msg":"skipping repo clone since no project was modified","json":{"repo":"<redacted>","pull":"38830"}}
```

</details>

Ulminator commented:
Hi @jamengual, just bumping this thread in hopes this bug fix can get merged in!

GenPage commented:
@Ulminator Can you resolve the file conflict and we'll get this in the next patch release. Thanks

Ulminator commented:
@GenPage just letting you know that I fixed the merge conflicts!

Ulminator commented on server/events/vcs/gitea/client.go:
```diff
@@ -421,13 +421,11 @@ func (c *GiteaClient) GetTeamNamesForUser(repo models.Repo, user models.User) ([
-func (c *GiteaClient) GetFileContent(logger logging.SimpleLogging, pull models.PullRequest, fileName string) (bool, []byte, error) {
-	logger.Debug("Getting file content for %s in Gitea pull request %d", fileName, pull.Num)
-
-	content, resp, err := c.giteaClient.GetContents(pull.BaseRepo.Owner, pull.BaseRepo.Name, pull.HeadCommit, fileName)
```
With this change removing the `models.PullRequest` from the interface I opted to changing the head commit here to the branch name which seems to be supported by the docs.

https://pkg.go.dev/code.gitea.io/sdk/gitea#Client.GetContents

https://gitea.com/gitea/go-sdk/src/commit/8b855b4a5b33bc2050143edc4044ee330bf09b05/gitea/repo_file.go#L196

[The name of the commit/branch/tag. Default the repository’s default branch (usually master)](https://docs.gitea.com/api/1.22/#tag/repository/operation/repoGetContentsList)

Ulminator commented:
@runatlantis/maintainers I know this has been stagnant for awhile, but I'd really like to get this merged. Since I originally wrote this PR the `gittea` client was added so I made updates there as well. PTAL and let me know if there's anything I need to do. 🙏

X-Guardian commented on server/events/project_command_builder.go:
```diff
@@ -347,7 +347,17 @@ func (p *DefaultProjectCommandBuilder) buildAllCommandsByCfg(ctx *command.Contex
 	if p.SkipCloneNoChanges && p.VCSClient.SupportsSingleFileDownload(ctx.Pull.BaseRepo) {
 		repoCfgFile := p.GlobalCfg.RepoConfigFile(ctx.Pull.BaseRepo.ID())
-		hasRepoCfg, repoCfgData, err := p.VCSClient.GetFileContent(ctx.Log, ctx.Pull, repoCfgFile)
+
```
Can you add a test for this new code?

Ulminator replied:
I just added a test case, regenerated the `mock_client.go` file with pegomock, and fixed the issue w/ the DCO status check on my first commit. Should be good for another look 👍

Ulminator replied:
I just added a test case, regenerated the mock_client.go file with pegomock, and fixed the issue w/ the DCO status check on my first commit.

The test also made it clear I could simplify some of the logic to just always pass the `HeadRepo` to `GetFileContent`. With that being the case, maybe I should just revert the interface modifications and add the `HeadRepo` to `models.PullRequest`? Not entirely sure how feasible that is

https://github.com/runatlantis/atlantis/blob/22b37512ae4205df901e8b8a3746a8ada4b52c50/server/events/models/models.go#L156-L181

jamengual commented:
@X-Guardian could you look at the latest updates? Thanks.

Ulminator commented:
@jamengual thank you for taking a look!

@X-Guardian please let me know if there are any further changes that need to be made!

X-Guardian commented on server/events/vcs/github_client.go:
```diff
@@ -1072,12 +1072,12 @@ func (g *GithubClient) ExchangeCode(logger logging.SimpleLogging, code string) (
 // The first return value indicates whether the repo contains a file or not
 // if BaseRepo had a file, its content will placed on the second return value
-func (g *GithubClient) GetFileContent(logger logging.SimpleLogging, pull models.PullRequest, fileName string) (bool, []byte, error) {
-	logger.Debug("Getting file content for %s in GitHub pull request %d", fileName, pull.Num)
```
Can you restore this debug message that you have removed?

Ulminator replied:
Since `pull models.PullRequest` is not being passed in anymore I do not have access to `pull.Num`, which is why I added this debug log in `project_command_builder.go`:

https://github.com/runatlantis/atlantis/blob/9b981f9b030d3a1dca7407731d9e2d6e1ea65158/server/events/project_command_builder.go#L351-L352

I did update the log though to just have the `repoCfgFile` and the `ctx.Pull.Num`, instead of just the full `ctx.Pull` struct.

X-Guardian replied:
The logger should already have the repo and pull request context set, so I would recommend leaving the debug message where it is, but just changing it to match the GitLab one, i.e.

```
logger.Debug("Getting GitHub file content for file '%s'", fileName)
```

X-Guardian replied:
Thanks for updating the PR description. Can you also update the PR title to better reflect the change (This is used for the changelog entry when it is released)

X-Guardian reviewed (changes requested):
Hi @Ulminator, just a small request to restore a debug line.

Can you also update the PR description, as the change doesn't actually check for a forked PR, it just uses the `headRepo` references from the context, rather than the `baseRepo`

Ulminator commented:
@X-Guardian & @jamengual 

PTAL when possible!

Had to resolve some merge conflicts as there were recent changes to `project_command_builder.go`, which ended up making the prior approval stale.

X-Guardian commented:
Hi @Ulminator, can you resolve the conflict in the test file?

Ulminator commented:
Hey @X-Guardian & @jamengual !

Apologies for taking so long to come back around to this PR. I have addressed all comments and this should be good for another look + hopefully get merged.

I see that the DCO check is failing, but it seems to be related to a commit you made @jamengual so I've decided to leave that as is.

Please let me know if you need anything else!

lukemassa commented:
LGTM, just make sure to add signoff to your commits, it looks like DCO is failing https://github.com/runatlantis/atlantis/pull/3900/checks?check_run_id=53298787867

jamengual commented:
Thanks @Ulminator for the contributor
